controllers/cruisecontroloperation_controller.go

```diff
@@ -0,0 +1,446 @@
+	case len(ccOperationQueueMap[ccOperationFirstExecution]) > 0:
+		ccOperationExecution = ccOperationQueueMap[ccOperationFirstExecution][0]
+	}
+
+	return ccOperationExecution
+}
+
+// SetupCruiseControlWithManager registers cruise control controller to the manager
+func SetupCruiseControlOperationWithManager(mgr ctrl.Manager) *ctrl.Builder {
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&banzaiv1alpha1.CruiseControlOperation{}).
+		WithEventFilter(SkipClusterRegistryOwnedResourcePredicate{}).
+		Named("CruiseControlOperation")
+
+	builder.WithEventFilter(
+		predicate.Funcs{
+			// We dont reconcile when there is no operation definied
+			CreateFunc: func(e event.CreateEvent) bool {
+				obj := e.Object.(*banzaiv1alpha1.CruiseControlOperation)
+				return obj.GetCurrentTaskOp() != ""
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldObj := e.ObjectOld.(*banzaiv1alpha1.CruiseControlOperation)
+				newObj := e.ObjectNew.(*banzaiv1alpha1.CruiseControlOperation)
+				if !reflect.DeepEqual(oldObj.GetCurrentTask(), newObj.GetCurrentTask()) ||
+					oldObj.GetDeletionTimestamp() != newObj.GetDeletionTimestamp() ||
+					oldObj.GetGeneration() != newObj.GetGeneration() {
+					return true
+				}
+				return false
+			},
+		})
+
+	return builder
+}
+
+func getKafkaClusterReference(operation *banzaiv1alpha1.CruiseControlOperation) (client.ObjectKey, error) {
+	if operation.GetClusterRef() == "" {
+		return client.ObjectKey{}, errors.NewWithDetails("could not find kafka cluster reference label for CruiseControlOperation", "missing label", banzaiv1beta1.KafkaCRLabelKey, "name", operation.GetName(), "namespace", operation.GetNamespace())
+	}
+	return client.ObjectKey{
+		Name:      operation.GetClusterRef(),
+		Namespace: operation.GetNamespace(),
+	}, nil
+}
+
+func updateResult(res *scale.Result, operation *banzaiv1alpha1.CruiseControlOperation, isAfterExecution bool) error {
+	if res == nil || operation.GetCurrentTask() == nil {
+		return nil
+	}
+	task := operation.GetCurrentTask()
+
+	if (task.State == banzaiv1beta1.CruiseControlTaskCompleted || task.State == banzaiv1beta1.CruiseControlTaskCompletedWithError) && task.Finished == nil {
+		task.Finished = &v1.Time{Time: time.Now()}
+	}
+
+	// Add the failed task into the status.failedTasks slice only when the update is happened after executing task
+	if isAfterExecution && task.Finished != nil && task.State == banzaiv1beta1.CruiseControlTaskCompletedWithError {
+		if len(operation.Status.FailedTasks) >= defaultFailedTasksHistoryMaxLength {
+			operation.Status.FailedTasks = append(operation.Status.FailedTasks[1:], *task)
+		} else {
+			operation.Status.FailedTasks = append(operation.Status.FailedTasks, *task)
+		}
+		operation.Status.NumberOfRetries += 1
+		operation.Status.ErrorPolicy = banzaiv1alpha1.ErrorPolicyRetry
+		task.Finished = nil
+		task.State = ""
+		task.Started = nil
+	}
+
+	task.ID = res.TaskID
+	if task.Started == nil {
+		//Sat, 27 Aug 2022 12:22:21 GMT
```

Comment:
Is this an example? If it is could we write `Example:` before it? Also a space should be added after the //.

Reply:
This can be removed.